trace/backend.go

```diff
@@ -0,0 +1,231 @@
+	addr           net.Addr
+	backoff        time.Duration
+	maxBackoff     time.Duration
+	connectTimeout time.Duration
+	bufferSize     uint
+	cap            uint
+}
+
+func (p *backendParams) params() *backendParams {
+	return p
+}
+
+// backend is a structure that can send an SSF span to a destination
+// over a persistent connection, handling reconnections. When
+// encountering connection errors, a backend will automatically attempt
+// to reconnect and blocks until reconnecting succeeds.
+//
+// Data loss / resiliency to failure
+//
+// If a backend encounters an error sending a span, it discards
+// the span and attempts to reconnect. This means that the
+// backend is resilient against "poison pill" spans, but it
+// also means that spans will get lost if there are connection
+// problems, such as veneurs getting restarted.
+type backend interface {
+	io.Closer
+
+	params() *backendParams
+	connection(net.Conn)
+
+	// sendSync synchronously sends a span to an upstream veneur. If
+	// it encounters a protocol error in sending, it loops
+	// forever, backing off by n*the backoff interval (until it
+	// reaches the maximal backoff interval) and tries to
+	// reconnect. If sendSync encounters any non-protocol errors
+	// (e.g. in serializing the SSF span), it returns them without
+	// reconnecting.
+	sendSync(ctx context.Context, span *ssf.SSFSpan) error
+
+	// flushSync causes all (potentially) buffered data to be sent to
+	// the upstream veneur.
+	flushSync(ctx context.Context) error
+}
+
+// packetBackend represents a UDP connection to a veneur server. It
+// does no buffering.
+type packetBackend struct {
+	backendParams
+	conn net.Conn
+}
+
+func (s *packetBackend) connection(conn net.Conn) {
+	s.conn = conn
+}
+
+func (s *packetBackend) Close() error {
+	if s.conn == nil {
+		return nil
+	}
+	return s.conn.Close()
+}
+
+func (s *packetBackend) sendSync(ctx context.Context, span *ssf.SSFSpan) error {
+	if s.conn == nil {
+		if err := connect(ctx, s); err != nil {
+			return err
+		}
+	}
+
+	data, err := proto.Marshal(span)
+	if err != nil {
+		return err
+	}
+	_, err = s.conn.Write(data)
+	return err
+}
+
+// flushSync on a PacketStream is a no-op.
+func (s *packetBackend) flushSync(context.Context) error {
+	return nil
+}
+
+var _ backend = &packetBackend{}
+
+// streamBackend is a backend for streaming connections.
+type streamBackend struct {
+	backendParams
+	conn   net.Conn
+	output io.Writer
+	buffer *bufio.Writer
+}
+
+func connect(ctx context.Context, s backend) error {
+	dialer := net.Dialer{}
+
+	params := s.params()
+	backoff := params.backoff
+	if backoff == 0 {
+		backoff = DefaultBackoff
+	}
+
+	maxBackoff := params.maxBackoff
+	if maxBackoff == 0 {
+		maxBackoff = DefaultMaxBackoff
+	}
+
+	connectTimeout := params.connectTimeout
+	if connectTimeout == 0 {
+		connectTimeout = DefaultConnectTimeout
+	}
+	ctx, cancel := context.WithTimeout(ctx, connectTimeout)
+	defer cancel()
+
+	wait := 0 * time.Second
```

Comment:
nit: `var wait time.Duration` is a little clearer to me, because `Second` isn't actually part of the the type (and the units don't matter on the fixed point anyway).